### PR TITLE
cmake/modules: Add zephyr_link_options (and friends)

### PR DESCRIPTION
**Pull request by keith-packard**
Cmake provides both target_link_libraries and target_link_options to separate linker options from dependent libraries. Mirror that by creating zephyr_link_options, then add zephyr_link_options_ifdef, zephyr_link_options_ifndef and target_link_options_ifdef convenience functions.

### CODE REVIEW

**Review by tejlmand — DISMISSED** [review]

**Comment by keith-packard** [comment]
Needed to also add `target_ld_libraries` as the POSIX code was using `target_ld_options` with libraries.

**Review by aescolar — DISMISSED** [review]
Ok for the native targets/ARCH_POSIX changes
(ok on a quick check for the rest)

File: cmake/modules/extensions.cmake
Comment (tejlmand) [inline cmake/modules/extensions.cmake:2101]: the existing `target_ld_options()` tries to verify that an option passed to the linker is indeed valid by the compiler (which invokes the linker).

That make sense for linker flags, but it doesn't make sense for a `-l<library>`, cause the `-l` itself is always valid, but whether the lib can be linked in depends on the library paths. 

So why this extra check if creating a dedicated `target_ld_libraries()` function ?

In addition, if we allow linking of libraries this way, then I guess it makes sense to omit the `-l` from the caller and let this function append `-l`.

That will even prevent callers from misusing `target_ld_libraries()` to pass linker options.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: cmake/modules/extensions.cmake
Comment (keith-packard) [reply cmake/modules/extensions.cmake:2101]: Oh, I assumed the check would also verify that the library existed, like `-ldl` as used in the posix backend?
Otherwise, yes, having it add `-l` to every option passed would avoid having it misused.

Comment (keith-packard) [reply cmake/modules/extensions.cmake:2101]: mostly, I didn't want to change *how* the options were validated, just the function used to get them into the link command

**Review by cfriedt — DISMISSED** [review]
LGTM

File: cmake/modules/extensions.cmake
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2191]: > Oh, I assumed the check would also verify that the library existed

how can it do that if it doesn't have all library path specified ?
Sure for libs available in default path, this would be possible, for other libs not in path, then this is not working.

At this stages we want to check if a linker flag (`-l`) is valid, not if the filibrary can be found.
The linking stage itself we provide info on missing libraries for linking.

Comment (keith-packard) [reply cmake/modules/extensions.cmake:2191]: that makes sense. And, we assume that the `-l` flag is valid because it's a linker. Sounds like it's a much simpler process in this case.

**Comment by keith-packard** [comment]
rebased and fixed a merge conflict

**Review by cfriedt — DISMISSED** [review]

**Comment by keith-packard** [comment]
I've tagged this for 3.7 so it's clear we don't need this for 3.6 (not that anyone should be looking at it, but just in case they are :-)

File: cmake/modules/extensions.cmake
Comment (tejlmand) [inline cmake/modules/extensions.cmake:2211]: This code part makes no sense now.

The NO_SPLIT is used in `target_ld_options()` to ensure that options are tested together in the `zephyr_check_compiler_flag()` call by setting `CMAKE_REQUIRED_FLAGS` to contain all options prior to that call.

As we are no longer calling `zephyr_check_compiler_flag()` then the logic of `NO_SPLIT` is not needed.

On top of that, CMake ensures that a call like `target_link_libraries(foo)` is translated in `-lfoo`*, meaning there is also need for the `JOIN .. "-l"` part.

This means this function can simply be written as:
```
function(target_ld_libraries target scope)
  target_link_libraries(${target} ${scope} ${ARGN})
endfunction()
```

Thus, the only reason for having `target_ld_libraries()` is to make a function similar to `target_ld_options()` but with a distinct name.

Also, it allows to extend the function later, if need be.

\* = **A plain library name:** The generated link line will ask the linker to search for the library (e.g. foo becomes -lfoo or foo.lib), ref: https://cmake.org/cmake/help/latest/command/target_link_libraries.html#overview.

**Review by tejlmand — CHANGES_REQUESTED** [review]
the function `target_ld_libraries()` can be reduced even further.

File: cmake/modules/extensions.cmake
Comment (keith-packard) [reply cmake/modules/extensions.cmake:2211]: Yeah, the addition of `-l` serves to ensure that this function can only be used to add libraries and not other link options. If you don't think that's a useful developer aid, I'll remove it. Otherwise, yup, this function is getting pretty small :-)

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Pull request closed without merging**
